pkg/rancher-desktop/main/extensions/manager.ts

```diff
@@ -233,56 +236,75 @@ class ExtensionManagerImpl implements ExtensionManager {
     return Buffer.from(origin.hostname, 'hex').toString();
   }
 
-  /**
-   * Convert incoming spawn options from the host-exec context.
-   */
-  protected convertHostOptions(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): SpawnOptions {
+  /** Spawn a process in the host context. */
+  protected spawnHost(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): ReadableChildProcess {
     const extensionId = this.getExtensionIdFromEvent(event);
     const extension = this.getExtension(extensionId) as ExtensionImpl;
-    const exePath = path.join(extension.dir, 'bin', options.command[0]);
 
-    return {
-      ...options,
-      command: [exePath, ...options.command.slice(1)],
-    };
+    if (!extension) {
+      throw new Error(`Could not find calling extension ${ extensionId }`);
+    }
+
+    return spawn(
+      path.join(extension.dir, 'bin', options.command[0]),
+      options.command.slice(1),
+      {
+        stdio: ['ignore', 'pipe', 'pipe'],
+        ..._.pick(options, ['cwd', 'env']),
+      });
   }
 
-  /**
-   * Convert incoming spawn options from the docker-exec context.
-   */
-  protected convertDockerCliOptions(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): SpawnOptions {
-    return {
-      ...options,
-      command: [this.client.executable, ...options.command],
-    };
+  /** Spawn a process in the docker-cli context. */
+  protected spawnDockerCli(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): ReadableChildProcess {
+    const extensionId = this.getExtensionIdFromEvent(event);
+    const extension = this.getExtension(extensionId) as ExtensionImpl;
+
+    if (!extension) {
+      throw new Error(`Could not find calling extension ${ extensionId }`);
+    }
+
+    return spawn(this.client.executable, options.command, {
+      stdio: ['ignore', 'pipe', 'pipe'],
+      ..._.pick(options, ['cwd', 'env']),
+    });
   }
 
-  /**
-   * Convert incoming spawn options from the container-exec context.
-   */
-  protected convertContainerOptions(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): Promise<SpawnOptions> {
+  /** Spawn a process in the container context. */
+  protected spawnContainer(event: IpcMainEvent | IpcMainInvokeEvent, options: SpawnOptions): Promise<ReadableChildProcess> {
     return Promise.reject(new Error('not implemented'));
   }
 
   /**
-   * Spawn a process on behalf of an extension, returning a promise that will be
-   * resolved when the process completes.
+   * Execute a process on behalf of an extension, returning a promise that will
+   * be resolved when the process completes.
    */
-  protected spawnBlocking(options: SpawnOptions): Promise<SpawnResult> {
-    const args = options.command.concat();
-    const exePath = args.shift();
+  protected execBlocking(process: ReadableChildProcess): Promise<SpawnResult> {
```

Comment:
The term "exec" is a little confusing in both `execBlocking` and `execStreaming` IMO. It doesn't match the meaning of "exec" in the nodejs context (i.e. short for "execute" AFAICT), and it doesn't match the meaning of "exec" in the Unix context. How about `handleBlocking` and `handleStreaming`? Or `manageProcessBlocking` and `manageProcessStreaming`? Some other verbs that might work are `administer`, `oversee`, `supervise`, and `conclude`.

Reply:
I was using it in the "execute a process" sense, but I can see how that's confusing. I think my preference is, in order:
- `processStreaming`
  - caveat: the thing _being_ processed is a process…
- `transformStreaming`
- `runStreaming`
- `handleStreaming`

Reply:
I'd get confused by`processStreaming` - as you say, the thing being processed is a process.

`transformStreaming` doesn't make sense to me - we aren't really transforming anything are we?

`runStreaming`: the process is already running isn't it?

Which leaves `handleStreaming`... which might imply the function being a handler for a `Streaming` or `Blocking` "event"?

I don't want to waste time bikeshedding though. It isn't that bad as it is. It's your call :smiley:

Reply:
Since there's no clear winner, let's stick with it as-is and rename once we have a clear winner?

Reply:
Sounds good to me.